src/user_parsed_input.rs

```diff
@@ -0,0 +1,284 @@
+    ///
+    /// # Panics
+    /// This function assume that Args and AppConfig are verfied eariler and are logicly correct
+    /// For example if both `--git` and `--path` are set this function will panic
+    pub fn try_from_args_and_config(app_config: &AppConfig, args: &Args) -> Self {
+        // --git and --path can not be set at the same time
+        assert!(args.git.is_none() || args.path.is_none());
+        let mut default_values = app_config.values.clone().unwrap_or_default();
+        let ssh_identity = app_config
+            .defaults
+            .as_ref()
+            .and_then(|dcfg| dcfg.ssh_identity.clone())
+            .or_else(|| args.ssh_identity.clone());
+
+        // --git
+        if let Some(git_url) = &args.git {
+            assert!(args.subfolder.is_none());
+
+            let git_user_in = GitUserInput::new(
+                git_url,
+                args.branch.clone(),
+                ssh_identity,
+                args.force_git_init,
+            );
+            return Self::new(git_user_in, args.favorite.clone(), default_values);
+        }
+
+        // --path
+        if let Some(path) = &args.path {
+            assert!(args.subfolder.is_none());
+
+            return Self::new(path, args.favorite.clone(), default_values);
+        }
+
+        // check if favorite is favorite configuration
+        assert!(args.favorite.is_some());
+        let fav_name = args.favorite.as_ref().unwrap();
+
+        if let Some(fav_cfg) = app_config.get_favorite_cfg(fav_name) {
+            assert!(fav_cfg.git.is_none() || fav_cfg.path.is_none());
+
+            let temp_location = if let Some(git_url) = &fav_cfg.git {
+                let branch = args.branch.clone().or_else(|| fav_cfg.branch.clone());
+                let git_user_input =
+                    GitUserInput::new(git_url, branch, ssh_identity, args.force_git_init);
+
+                TemplateLocation::from(git_user_input)
+            } else if let Some(path) = &fav_cfg.path {
+                TemplateLocation::from(path)
+            } else {
+                unreachable!();
+            };
+
+            if let Some(fav_default_values) = &fav_cfg.values {
+                default_values.extend(fav_default_values.clone());
+            }
+
+            return Self::new(
+                temp_location,
+                args.subfolder.clone().or_else(|| fav_cfg.subfolder.clone()),
+                default_values,
+            );
+        }
+
+        // there is no specified favorite in configuration
+        // this part try to guess what user wanted in order:
+        // 1. look for abbrevations like gh:, gl: etc.
+        // 2. check if template directory exist
+        // 3. check if the input is in form org/repo<> (map to github)
+        // 4. assume user wanted use --git
+
+        let temp_location = if let Some(git_url) = abbreviated_git_url_to_full_remote(fav_name) {
```

Comment:
could you please look at the clippy lint:

```
error: use Option::map_or_else instead of an if let/else
   --> src/user_parsed_input.rs:115:29
    |
115 |           let temp_location = if let Some(git_url) = abbreviated_git_url_to_full_remote(fav_name) {
    |  _____________________________^
116 | |             TemplateLocation::from(GitUserInput::with_git_url_and_args(&git_url, args))
117 | |         } else if let Some(path) = local_path(fav_name) {
118 | |             TemplateLocation::Path(path)
...   |
127 | |             ))
128 | |         };
    | |_________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#option_if_let_else
help: try
    |
115 ~         let temp_location = abbreviated_git_url_to_full_remote(fav_name).map_or_else(|| if let Some(path) = local_path(fav_name) {
116 +             TemplateLocation::Path(path)
117 +         } else if let Some(git_url) = abbreviated_github(fav_name) {
118 +             TemplateLocation::from(GitUserInput::with_git_url_and_args(&git_url, args))
119 +         } else {
120 +             TemplateLocation::from(GitUserInput::new(
  ...
```

Reply:
This is a little bit complicated. This is what clippy suggested:

```rust
let temp_location = abbreviated_git_url_to_full_remote(fav_name).map_or_else(
    || {
        local_path(fav_name).map_or_else(
            || {
                abbreviated_github(fav_name).map_or_else(
                    || {
                        TemplateLocation::from(GitUserInput::new(
                            fav_name,
                            args.branch.clone(),
                            ssh_identity,
                            args.force_git_init,
                        ))
                    },
                    |git_url| {
                        TemplateLocation::from(GitUserInput::with_git_url_and_args(
                            &git_url, args,
                        ))
                    },
                )
            },
            TemplateLocation::Path,
        )
    },
    |git_url| TemplateLocation::from(GitUserInput::with_git_url_and_args(&git_url, args)),
)
```
I will refactor it to make clippy happy but still not make this so nested.